src/rust/engine/serverset/src/lib.rs

```diff
@@ -0,0 +1,373 @@
+}
+
+#[derive(Clone, Copy, Debug)]
+pub enum Health {
+  Healthy,
+  Unhealthy,
+}
+
+#[derive(Debug)]
+struct UnhealthyInfo {
+  unhealthy_since: Instant,
+  next_attempt_after: Duration,
+}
+
+#[derive(Debug)]
+struct Backend<T> {
+  server: T,
+  unhealthy_info: Arc<Mutex<Option<UnhealthyInfo>>>,
+}
+
+#[derive(Clone, Copy, Debug)]
+pub struct BackoffConfig {
+  ///
+  /// The time a backend will be skipped after it is first reported unhealthy.
+  ///
+  pub initial_lame: Duration,
+  ///
+  /// Ratio by which to multiply the most recent lame duration if a backend continues to be
+  /// unhealthy.
+  ///
+  /// The inverse is used when easing back in after health recovery.
+  pub backoff_ratio: f64,
+
+  ///
+  /// Maximum duration to wait between attempts.
+  ///
+  pub max_lame: Duration,
+}
+
+impl<T: Clone + Send + Sync + 'static> Serverset<T> {
+  pub fn new(servers: Vec<T>, backoff_config: BackoffConfig) -> Result<Self, String> {
+    if servers.is_empty() {
+      return Err("Must supply some servers".to_owned());
+    }
+
+    let BackoffConfig {
+      initial_lame,
+      backoff_ratio,
+      max_lame,
+    } = backoff_config;
+
+    if backoff_ratio < 1.0 {
+      return Err(format!(
+        "Failure backoff ratio must be at least 1, got: {}",
+        backoff_ratio
+      ));
+    }
+    let backoff_ratio =
+      num_rational::Ratio::<i8>::approximate_float(backoff_ratio).ok_or_else(|| {
+        format!(
+          "Couldn't find reasonable backoff ratio for {}",
+          backoff_ratio
+        )
+      })?;
+    let backoff_ratio =
+      num_rational::Ratio::new(*backoff_ratio.numer() as u32, *backoff_ratio.denom() as u32);
+    Ok(Serverset {
+      inner: Arc::new(Inner {
+        servers: servers
+          .into_iter()
+          .map(|s| Backend {
+            server: s,
+            unhealthy_info: Arc::new(Mutex::new(None)),
+          }).collect(),
+        next: AtomicUsize::new(0),
+        failure_initial_lame: initial_lame,
+        failure_backoff_ratio: backoff_ratio,
+        failure_max_lame: max_lame,
+      }),
+    })
+  }
+
+  ///
+  /// Get the next (probably) healthy backend to use.
+  ///
+  /// The caller will be given a backend to use, and should call the supplied callback with the
+  /// observed health of that backend.
+  ///
+  /// If the callback is not called, the health status of the server will not be changed from its
+  /// last known status.
+  ///
+  /// If all resources are unhealthy, this function will block the calling thread until the backoff
+  /// period has completed. We'd probably prefer to use some Future-based scheduling, but that
+  /// would require this type to be Resettable because of our fork model, which would be very
+  /// complex.
+  ///
+  pub fn next(&self) -> (T, Box<Fn(Health) + Send + Sync>) {
+    let (i, server) = loop {
+      let i = self.inner.next.fetch_add(1, Ordering::Relaxed) % self.inner.servers.len();
+      let server = &self.inner.servers[i];
```

Comment:
I assume that `inner.servers` only gets mutated at creation and the privateness of that field guarantees that it doesn't get mutated elsewhere. If that is the case, this is okay. Otherwise, there could be a disconnect between `i` and the actual vector.

Reply:
Still maybe worth throwing an RWLock, if reading has little overhead.

Reply:
In fact, the fact that we `Arc` up the `Inner` means that the `Vec` _cannot_ ever be mutated, because there's no way for anyone to get a mutable reference to it (unless they are the only holder of the `Arc`, in which case they would need to unwrap it, which proves that no one else could have any dangling callbacks) :) Yay rust!